src/memory_prefetch.c

```diff
@@ -8,15 +8,282 @@
  */
 
 #include "memory_prefetch.h"
+#include "server.h"
+
+typedef enum {
+    PREFETCH_ENTRY, /* Initial state, prefetch entries associated with the given key's hash */
+    PREFETCH_VALUE, /* prefetch the value object of the entry found in the previous step */
+    PREFETCH_DONE   /* Indicates that prefetching for this key is complete */
+} PrefetchState;
+
+typedef struct KeyPrefetchInfo {
+    PrefetchState state; /* Current state of the prefetch operation */
+    hashtableIncrementalFindState hashtab_state;
+} KeyPrefetchInfo;
+
+/* PrefetchCommandsBatch structure holds the state of the current batch of client commands being processed. */
+typedef struct PrefetchCommandsBatch {
+    size_t cur_idx;                 /* Index of the current key being processed */
+    size_t keys_done;               /* Number of keys that have been prefetched */
+    size_t key_count;               /* Number of keys in the current batch */
+    size_t client_count;            /* Number of clients in the current batch */
+    size_t max_prefetch_size;       /* Maximum number of keys to prefetch in a batch */
+    size_t executed_commands;       /* Number of commands executed in the current batch */
+    int *slots;                     /* Array of slots for each key */
+    void **keys;                    /* Array of keys to prefetch in the current batch */
+    client **clients;               /* Array of clients in the current batch */
+    hashtable **keys_tables;        /* Main table for each key */
+    hashtable **expire_tables;      /* Expire table for each key */
+    KeyPrefetchInfo *prefetch_info; /* Prefetch info for each key */
+} PrefetchCommandsBatch;
+
+static PrefetchCommandsBatch *batch = NULL;
```

Comment:
high level comment:
I think this is fine that we have a single global batch for reading jobs from io-threads. I do think that we could have all the related functions accept a batch as input so it might be used for other future cases were we would maybe like to start a batch on the engine side for some other purpose.
So we could have a  
static PrefetchCommandsBatch *g_batch = NULL;
and pass it to the relevant functions.

Reply:
Agree this can be improved, but let's consider the size of the hashstable feature diff. It's already quite huge. This code is the same as in unstable. I checked the diff between this file in unstable locally and it really keeps what can be kept from unstable.